test/NuGet.Core.Tests/NuGet.Commands.Test/BuildAssetsUtilsTests.cs

```diff
@@ -19,6 +19,7 @@
 
 namespace NuGet.Commands.Test
 {
+    [Collection(nameof(NotThreadSafeResourceCollection))]
     public class BuildAssetsUtilsTests
```

Comment:
Why do we think this can't run in parallel?

Reply:
`Tests_On_Linux` job hung/failed in [build](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4113212&view=logs&j=bd7d45da-bd9f-59a9-6834-84d44942bc5e) because `NuGet.Commands.Test` project had around twenty tests that were long running. I am sure there was a test in `BuildAssetsUtilsTests` reported as long running by `xUnit`. Unfortunately, I see [blank screen](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4113212&view=logs&j=bd7d45da-bd9f-59a9-6834-84d44942bc5e&t=d80d8349-9076-5823-9e7a-95c7c13ef00c) in the build logs when I tried to find the test that was reported as `long running` by `xUnit`.

Reply:
I created a new branch on top of `dev` and then modified [`xunit settings to detect long running tests`](https://github.com/NuGet/NuGet.Client/commit/c310c818e95177210f6132c1e2d955ff5990442d). I have observed that `Tests On Linux` failed in [build](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=4123010&view=results) due to long running/hung tests in `NuGet.Protoco/Testsl`. However, there are 9 tests different from previous set of long running tests. The reason I marked only few test classes single threaded is to not increase CI duration. If different tests are reported as long running in each build, then we may still have hangs in CI. Looking at this pattern, I am considering of running tests in `NuGet.Protocol.Tests` and `NuGet.Commands.Tests` projects single threaded. This approach will hopefully avoid CI hangs in Linux. We can reenable parallelism for these projects once we have implemented dump collector.

@zivkan - Pls let me know your feedback.

Reply:
At this time I'm not so concerned about long running tests if they recover/finish, but I'm very concerned about hanging tests.

If tests are hanging because of some kind of shared resource issue when the tests run in parallel, then making those tests run sequentially will help the tests complete, but will "hide" a bug (which I hope is a bug in our test code, not product code). On the other hand, xunit uses task synchronization contexts, and if our code has a timing bug that occasionally results in a "main thread" deadlock, it doesn't matter if we run them in parallel or sequentially, the deadlock remains.

Having said that, [I wrote a log parser to determine which tests started and never finished](https://github.com/NuGet/Entropy/tree/master/client-ci-analysis/find-incomplete-tests), and I found that every single time it happens, the number of tests that didn't complete is exactly equal to the CI agent's `Environment.ProcessorCount`. This means that either xunit's synchronization context does not work like a GUI app's UI thread (only 1), or there's something else at play. It makes me think running the tests sequentially could help, because two tests which block each other, or task-deadlock bugs should sometimes result in fewer tests hanging than the number of CPU cores. But we don't see this, only tests complete, or (CPU count) tests froze. 

I feel like this PR is just throwing a rock into the dark. We don't know why tests are unable to finish, and we're not even sure this will mitigate the problem. However, given that tests on linux are hanging more often than ever, and we've started getting some windows func test and unit test hangs, I think it's worth a try until we can implement the hung test memory dump to get real answers.

Reply:
I searched "Linux hang" in https://github.com/dotnet/runtime/issues, there're several open issues. 
Not sure if they're related to our test hangs.